docs/proposals/20230706-yurtappconfigurationreplacement.md

```diff
@@ -0,0 +1,291 @@
+type ImageItem struct {
+	// ContainerName represents name of the container 
+	// in which the Image will be replaced
+	ContainerName string `json:"containerName"`
+	// ImageClaim represents the claimed image name 
+	// which is injected into the container above
+	ImageClaim string `json:"imageClaim"`
+}
+
+// EnvItem specifies the corresponding container and the claimed env
+type EnvItem struct {
+	// ContainerName represents name of the container 
+	// in which the env will be replaced
+	ContainerName string `json:"containerName"`
+	// EnvClaim represents the detailed enviroment variables container contains
+	EnvClaim map[string]string `json:"envClaim"`
+}
+
+// PersistentVolumeClaimItem specifies the corresponding container and the claimed pvc
+type PersistentVolumeClaimItem struct {
+	// PVCSource represents pvcClaim name.
+	PVCSource string `json:"pvcSource"`
+	// PVCTarget represents the PVC corresponding to the volume above.
+	// PVCTarget supprot advanced features like wildcard. 
+	// By naming pvc as pvcName-{{nodepool}}, all pvc can be injected at once.
+	PVCTarget string `json:"pvcTarget"`
+}
+
+// ConfigMapItem specifies the corresponding containerName and the claimed configMap
+type ConfigMapItem struct {
+	// ConfigMapSource represents configMap name
+	ConfigMapSource string `json:"configMapSource"`
+	// ConfigMapTarget represents the ConfigMap corresponding to the volume above. 
+	// ConfigMapTarget supprot advanced features like wildcard. 
+	// By naming configMap as configMapName-{{nodepool}}, all configMap can be injected at once.
+	ConfigMapTarget string `json:"configMapTarget"`
+}
+
+type SecretItem struct {
+	// SecretSource represents secret name.
+	SecretSource string `json:"secretSource"`
+	// SecretTarget represents the Secret corresponding to the volume above.
+	// SecretTarget supprot advanced features like wildcard. 
+	// By naming secret as secretName-{{nodepool}}, all secret can be injected at once.
+	SecretTarget string `json:"secretTarget"`
+}
+
+// Item represents configuration to be injected.
+// Only one of its members may be specified.
+type Item struct {
+	Image                 *ImageItem                 `json:"image"`
+	ConfigMap             *ConfigMapItem             `json:"configMap"`
+	Secret                *SecretItem                `json:"secret"`
+	Env                   *EnvItem                   `json:"env"`
+	PersistentVolumeClaim *PersistentVolumeClaimItem `json:"persistentVolumeClaim"`
+	Replicas              *int                       `json:"replicas"`
+	UpgradeStrategy       *string                    `json:"upgradeStrategy"`
+}
+
+// Replacement describe a set of nodepools and their shared or identical configurations
+type Replacement struct {
+	Pools []string `json:"pools"`
+	Items []Item   `json:"items"`
+}
+
+type Subject struct {
+	metav1.TypeMeta `json:",inline"`
+	// NameSpace is the namespace of YurtAppSet or YurtAppDaemon
+	NameSpace string `json:"nameSpace"`
```

Comment:
Can this be as consistent as possible with Metadata? I want this to be unified with the Namespace of k8s, so that we can call it more easily.
```go
// Namespace defines the space within which each name must be unique. An empty namespace is
// equivalent to the "default" namespace, but "default" is the canonical representation.
// Not all objects are required to be scoped to a namespace - the value of this field for
// those objects will be empty.
//
// Must be a DNS_LABEL.
// Cannot be updated.
// More info: http://kubernetes.io/docs/user-guide/namespaces
// +optional
Namespace string `json:"namespace,omitempty" protobuf:"bytes,3,opt,name=namespace"`
```

Reply:
You're right. It is more reasonable to use the Namespace of k8s.

Reply:
@vie-serendipity @rambohe-ch I think it's reasonable. While nodepool is cluster-level, the yurtappdaemons, yurtappsets, and yurtstaticsets that need to be modified are all namespace-level. If we assume YurtAppConfigurationReplacement is a namespace level of CRD, I don't think cross-namespace modification should be allowed.
By the way, we're not going to set it to cluster-level CRD, are we?

Reply:
@LavenderQAQ  It should be a namespace level CRD. I agree with @rambohe-ch , we can remove the namespace field in the Subject.

Reply:
@LavenderQAQ YurtAppConfigurationReplacement is used to handle namespace scoped resource(such as Deployment) in NodePool, so it is reasonable to define YurtAppConfigurationReplacement as namespace scoped resource.